src/cast.c

```diff
@@ -261,7 +263,111 @@ SEXP vec_cast(SEXP x, SEXP to) {
   return out;
 }
 
+// Copy attributes except names. This duplicates `x` if needed.
+SEXP vctrs_restore_default(SEXP x, SEXP to) {
+  int n_protect = 0;
+
+  SEXP attrib = PROTECT(Rf_shallow_duplicate(ATTRIB(to)));
+  ++n_protect;
+
+  if (attrib == R_NilValue) {
+    UNPROTECT(n_protect);
+    return x;
+  }
+
+  if (MAYBE_REFERENCED(x)) {
+    x = PROTECT(Rf_shallow_duplicate(x));
+    ++n_protect;
+  }
+
+  // Copy attributes but keep names
+  SEXP nms = Rf_getAttrib(x, R_NamesSymbol);
+  SET_ATTRIB(x, attrib);
+  Rf_setAttrib(x, R_NamesSymbol, nms);
+
+  // SET_ATTRIB() does not set object bit when attributes include class
+  if (OBJECT(to)) {
+    SET_OBJECT(x, true);
+  }
+
+  UNPROTECT(n_protect);
+  return x;
+}
+
+static R_len_t df_raw_length(SEXP x) {
+  SEXP attrib = ATTRIB(x);
+
+  // Avoid materialising compact sequences
+  while (attrib != R_NilValue) {
+    if (TAG(attrib) == R_RowNamesSymbol) {
+      SEXP rn = CAR(attrib);
+
+      if (is_compact_rownames(rn)) {
+        return compact_rownames_length(rn);
+      } else {
+        return Rf_length(rn);
+      }
+    }
+
+    attrib = CDR(attrib);
+  }
+
+  if (Rf_length(x) >= 1) {
+    return vec_size(VECTOR_ELT(x, 0));
+  } else {
+    return 0;
+  }
+}
+
+static SEXP df_restore(SEXP x, SEXP to) {
+  if (TYPEOF(x) != VECSXP) {
+    Rf_errorcall(R_NilValue, "Internal error: Attempt to restore data frame from a %s.",
+                 Rf_type2char(TYPEOF(x)));
+  }
+
+  int n_protect = 0;
+
+  // Compute size before changing attributes
+  R_len_t size = df_raw_length(x);
+
+  if (MAYBE_REFERENCED(x)) {
```

Comment:
This is an optimisation to avoid duplicating unnamed inputs?

Reply:
When `vctrs_restore()` is called from R we need to duplicate to avoid changing attributes by reference. But we will also call this method from C with a fresh list that we own, and then it doesn't need to be duplicated.